### DISCUSSION TITLE
Drop globby as a dependency

### DISCUSSION THREAD

**Discussion by PodaruDragos** [mikro-orm/mikro-orm#4837] [Ideas]
Hello,  as `v6` is just around the block please consider dropping `globby` as a dependency. 
Just recursively searching for files in a `dir` can be obtained without 23 dependencies. 

Example code ( tested locally on random code, not Mikro )

```
export function glob(
  pattern: RegExp,
  basePath: string
): string[] {
  const files: string[] = [];

  function traverseDirectory(directory: string): void {
    const entries = readdirSync(directory);

    for (const entry of entries) {
      const entryPath = join(directory, entry);
      const stat = statSync(entryPath);

      if (stat.isDirectory()) {
        traverseDirectory(entryPath);
      } else if (stat.isFile() && entry.match(pattern)) {
        files.push(entryPath);
      }
    }
  }

  traverseDirectory(basePath);
  return files;
}



const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const files = glob(/\.ts$/, join(__dirname, './entities'));
console.log({ files });

```

What do you think ?

**Comment by B4nan** [comment]
I can't say I like the dependency, but I have so many real problems to solve, that I am simply not willing to spend time on making a globby replacement myself. Feel free to PR that if you have the time and you consider this important for you - I am all in for a smaller dependency footprint. But I want something that is pretty much a drop-in replacement (I don't care about the API, feature-wise). We need globs, not regexps. We also need negative globs. And such a thing should be well tested on windows too (I can't do that myself and not really interested in slow CI windows tests either, the CI is already very slow with linux). 

But that is pretty much it, it's not that unreal to achieve this. It's just about the time and priorities, and I have way too much more important stuff right now.

Worth saying that if we make it a compatible replacement, it can happen any time, as it won't be breaking. That is what I'd envision ideally.

**Comment by PodaruDragos** [comment]
Fair enough, by looking at the code I can see globby being used in 3 places.
`FileCacheAdapter`, `MetadataDiscovery` and `SeedManager`. now there are no negating paths no where there, it's just getting files from a directory as far as I can tell. Is there any reason why you would want to support other things ?

I can test this on windows, since I am working from a windows machine.
In regards of globs, I am sure I can adapt that function to work with something like : `'/**/entities/*.ts'` and `'!/**/entities/*.ts'`.

I'll try to cook up something and hopefully if I can run the tests on my local machine I will make a PR.

**Reply by B4nan** [reply]
> now there are no negating paths no where there

This is user-facing, it's not about internal usage.

**Reply by B4nan** [reply]
Btw maybe we could use some lower-level dependency of globby to make things a bit simpler. 

And one more note, globs are generally more powerful than just handling `**`, e.g. there are things like `file-{a,b,c}.jpg` supported. We don't necessarily need to support those out of box, but if we won't, we should have an easy way to opt into the previous behavior (or bring the globby back in from outside).

**Reply by PodaruDragos** [reply]
well, I have tried a quick prototype. Writing a function with no dependencies that can also handle globs like this `!(*.d).{js,ts}` is complicated to say the least. I will spend some more time on this, but I think this is probably more complicated than I initially thought.

**Reply by B4nan** [reply]
Yeah but maybe we could just pick some of the lower level dependency (e.g. `picomatch`).

**Comment by PodaruDragos** [comment]
Hey Martin, good news

[glob pr](https://github.com/nodejs/node/pull/51912).

I mean it's still not there and probably `@types/node` won't have it for a long time but this is really good.

PS: it's using `minimatch` so it should support all usecases

**Reply by B4nan** [reply]
Would be great if it could make it into node v20 lts, as that will be probably the minimal supported version of the next ORM major version - but let's see, that's long way ahead of us (and I still envision next major as ESM only too, especially given the recent PR to node with support for `require(esm)`).

**Reply by PodaruDragos** [reply]
yeah. there is also now support to parseEnvFiles so that is also good news, no more `dotenv`

**Reply by PodaruDragos** [reply]
to be fair things are looking so good in all aspects, when v8 ships native decorators and decorators metadata and it's chery picked into node.js, `reflect-metadata` can go as well. 

If I am not mistaken, mikro does not use class parameter decorators right ?

**Reply by B4nan** [reply]
> If I am not mistaken, mikro does not use class parameter decorators right ?

nope, but maybe we should :D now to get the parameter names we need to parse the JS AST, which also brings one fat dependency on acorn (that slows down the discovery significantly, IIRC its literally the only real bottleneck). in the early versions, i was using regexps to parse things manually, as i didn't want to bring fat deps for such a small reason, but it was so broken that i moved to escaya and later on to acorn as escaya wasn't maintained (at all)

**Reply by PodaruDragos** [reply]
if you only want to get lets say we have this

```
@Entity()
export class SomeEntity {
  @Property()
  public property
}
```

you can read property with the new class based decorators, I am talking about vanilla decorators. I can post an example if you are interested. Ofc that would be a massive breaking change haha, but it is possible

**Reply by B4nan** [reply]
I am talking about `constructor` (and `toJSON`) parameters, the ORM needs to know their names when it invokes them. And we parse the AST to get those names currently.

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/Utils.ts#L474

(we actually use esprima nowadays, used to be acorn and escaya before)

**Reply by PodaruDragos** [reply]
`toJSON` can be read if it's decorated, with constructor is not possible if they don't advance the parameter decorators.

**Comment by B4nan** [comment]
FYI v7 will replace `globby` with `tinyglobby` https://github.com/mikro-orm/mikro-orm/commit/f6caf68fd569789094a9d4fb22133764d2b5126f

also tried to use the native glob from node, but that would print a warning about the feature being experimental, so I'd rather not use that (also it failed in one test case for some reason)

**Reply by PodaruDragos** [reply]
nice !!! thanks the update, weird that the node one fails in one test, it is based on `minimatch` which should have every case as `globby`. Experimental flag will soon go away as all the experimental flags get dropped in newer releases. does `v24` still list it as experimental ?

**Reply by B4nan** [reply]
try yourself to see 🙃 but even if that would be the case, its still the same problem, i don't want to use API that would print a warning for users on node 22 🤷 

note that it didn't fail as in a wrong behaviour would happen, it didn't work at all, the tests were broken completely and didn't report any failure, no error message whatsoever.

**Reply by PodaruDragos** [reply]
either way `tinyglobby` is much better than `globby` so thanks for taking the time to change it
